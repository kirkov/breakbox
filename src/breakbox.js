import React from 'react'
import cxs from 'cxs/monolithic'
import defaultConfig from './config'
import contextTypes from './context-types'

cxs.prefix('_bbx_')

const resolveValue = (spaces, value) => {
  if (typeof value !== 'number') return value

  return value <= spaces.length
    ? spaces[value - 1]
    : value + 'px'
}

const getStyles = (props, config) => {
  const { spaces } = config

  // Add '0' value for default value (without a media query)
  const breakpoints = [0, ...config.breakpoints]
  const stylesPerBreakpoint = breakpoints.map(b => [])

  Object.entries(props).forEach(([prop, values]) => {
    if (Array.isArray(values)) {
      for (let i = 0; i < values.length; i++) {
        const value = values[i]
        if (value === null || value === undefined) continue

        const resolvedValue = resolveValue(spaces, value)
        stylesPerBreakpoint[i].push([prop, resolvedValue])
      }
    } else if (values !== null && values !== undefined) {
      stylesPerBreakpoint[0].push([prop, resolveValue(spaces, values)])
    }
  })

  const styles = stylesPerBreakpoint.reduce((result, declarations, index) => {
    // const media = stylesPerBreakpoint.length-1 !== index
    //   ? `@media (min-width: ${breakpoints[(stylesPerBreakpoint.length-1) - index]})`
    //   : null

    const media = index !== 0
      ? `@media (min-width: ${breakpoints[index]})`
      : null

    const declarationsObj = declarations.reduce((res, declaration) => {
      const [property, value] = declaration
      res[property] = value
      return res
    }, {})

    if (media === null) result = { ...result, ...declarationsObj }
    else result[media] = declarationsObj

    return result
  }, {})

  Object.keys(styles).forEach(key => (Object.keys(styles[key]).length === 0) && delete styles[key])

  return styles
}

export const breakbox = (containerStyles = {}) => {
  const BreakboxComponent = ({className, children, tag = 'div', ...otherProps}, context) => {
    const config = Object.assign({}, defaultConfig, context.breakbox)
    const styles = getStyles(Object.assign(otherProps, containerStyles), config)
    const stylesClassName = cxs(styles)

    // used for testing
    let debug = null
    if (process.env.NODE_ENV === 'test') {
      debug = <style key='css' dangerouslySetInnerHTML={{ __html: cxs.css() }} />
    }

    return React.createElement(
      tag,
      {
        className: [className, stylesClassName].join(' ').trim()
      },
      [ children, debug ]
    )
  }

  BreakboxComponent.contextTypes = contextTypes

  return BreakboxComponent
}

export const resetStyles = () => cxs.reset()
