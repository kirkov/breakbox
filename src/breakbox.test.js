/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import {Provider, Container, Box} from '.'
import {resetStyles} from './breakbox'

const matchSnapshot = (name, rendered) => {
  // resetStyles()
  expect(renderer.create(rendered).toJSON()).toMatchSnapshot(name)
}

beforeEach(() => {
  // resetStyles()
})

it('can specify element tag', () => {
  matchSnapshot('container can render an article tag',
    <Container tag='article' background='red' />
  )
})

it('skips breakpoints', () => {
  matchSnapshot('skips middle value',
    <Container flexDirection={['row', null, 'column']} />
  )

  matchSnapshot('skips multiple values',
    <Container flexDirection={['row', null, 'column', null]} />
  )

  matchSnapshot('skips null',
    <Container flexDirection={[null]} />
  )

  matchSnapshot('skips undefined',
    <Container flexDirection={[undefined]} />
  )

  matchSnapshot('skips empty string',
    <Container flexDirection={['']} />
  )

  matchSnapshot('skips empty',
    <Container flexDirection={[]} />
  )
})

it('handle single values', () => {
  matchSnapshot('single value',
    <Container flexDirection='row' />
  )

  matchSnapshot('null value',
    <Container flexDirection={null} />
  )
})

it('can provide custom breakpoints', () => {
  matchSnapshot('uses custom breakpoints',
    <Provider config={{ breakpoints: ['100px', '200px', '300px'] }}>
      <Container width={[50, 150, 250]} />
    </Provider>
  )
})

it('can provide custom spaces', () => {
  matchSnapshot('uses custom spaces',
    <Provider config={{ spaces: ['100px', '200px', '300px'] }}>
      <Container marginRight={['50', '150', '250']} />
    </Provider>
  )

  matchSnapshot('looks up the custom space values',
    <Provider config={{ spaces: ['10px', '20px', '30px'] }}>
      <Box paddingRight={[1, 2, 3]} />
    </Provider>
  )

  matchSnapshot('handle unknowns space',
    <Provider config={{ spaces: ['10px', '20px'] }}>
      <Box paddingRight={[1, 2, 3]} />
    </Provider>
  )
})

// function Test () {
//   const myRef = React.useRef()
//   return (
//     <Provider spaces={['10px', '20px']}>
//       <Box ref={myRef} paddingRight={[1, 2, 3]} />
//     </Provider>
//   )
// }

// it.only('forwards ref', () => {



//   // console.log({ myRef })

//   const testRenderer = renderer.create(<Test />)

//   // console.log(myRef)

// })
