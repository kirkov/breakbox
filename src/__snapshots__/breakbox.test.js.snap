// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can provide custom breakpoints: uses custom breakpoints 1`] = `
<div
  className="_bbx_0 _bbx_1 _bbx_2"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{width:50px}@media (min-width: 100px){._bbx_1{width:150px}}@media (min-width: 200px){._bbx_2{width:250px}}",
      }
    }
  />
</div>
`;

exports[`can provide custom spaces: handle unknowns space 1`] = `
<div
  className="_bbx_0 _bbx_1 _bbx_2"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{padding-right:10px}@media (min-width: 600px){._bbx_1{padding-right:20px}}@media (min-width: 900px){._bbx_2{padding-right:3px}}",
      }
    }
  />
</div>
`;

exports[`can provide custom spaces: looks up the custom space values 1`] = `
<div
  className="_bbx_0 _bbx_1 _bbx_2"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{padding-right:10px}@media (min-width: 600px){._bbx_1{padding-right:20px}}@media (min-width: 900px){._bbx_2{padding-right:30px}}",
      }
    }
  />
</div>
`;

exports[`can provide custom spaces: uses custom spaces 1`] = `
<div
  className="_bbx_0 _bbx_1 _bbx_2"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{margin-right:50}@media (min-width: 600px){._bbx_1{margin-right:150}}@media (min-width: 900px){._bbx_2{margin-right:250}}",
      }
    }
  />
</div>
`;

exports[`can specify element tag: container can render an article tag 1`] = `
<article
  className="_bbx_0"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{background:red}",
      }
    }
  />
</article>
`;

exports[`handle single values: null value 1`] = `
<div
  className="_bbx_0"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{}",
      }
    }
  />
</div>
`;

exports[`handle single values: single value 1`] = `
<div
  className="_bbx_0"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{flex-direction:row}",
      }
    }
  />
</div>
`;

exports[`skips breakpoints: skips empty 1`] = `
<div
  className="_bbx_0"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{}",
      }
    }
  />
</div>
`;

exports[`skips breakpoints: skips empty string 1`] = `
<div
  className="_bbx_0"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{}",
      }
    }
  />
</div>
`;

exports[`skips breakpoints: skips middle value 1`] = `
<div
  className="_bbx_0 _bbx_1"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{flex-direction:row}@media (min-width: 900px){._bbx_1{flex-direction:column}}",
      }
    }
  />
</div>
`;

exports[`skips breakpoints: skips multiple values 1`] = `
<div
  className="_bbx_0 _bbx_1"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{flex-direction:row}@media (min-width: 900px){._bbx_1{flex-direction:column}}",
      }
    }
  />
</div>
`;

exports[`skips breakpoints: skips null 1`] = `
<div
  className="_bbx_0"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{}",
      }
    }
  />
</div>
`;

exports[`skips breakpoints: skips undefined 1`] = `
<div
  className="_bbx_0"
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "._bbx_0{}",
      }
    }
  />
</div>
`;
